src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NuGetInstaller.cs

```diff
@@ -0,0 +1,336 @@
+                }
+
+                //not a local package file
+                return Task.FromResult(validPackageId && hasValidVersion);
+            }
+            _environmentSettings.Host.LogDiagnosticMessage($"{installationRequest.PackageIdentifier} is identified as the local NuGet package.", DebugLogCategory);
+            return Task.FromResult(true);
+        }
+
+        public IManagedTemplatePackage Deserialize(IManagedTemplatePackageProvider provider, TemplatePackageData data)
+        {
+            _ = provider ?? throw new ArgumentNullException(nameof(provider));
+            if (data.InstallerId != Factory.Id)
+            {
+                throw new ArgumentException($"{nameof(NuGetInstaller)} can only deserialize packages with {nameof(data.InstallerId)} {Factory.Id}", nameof(data));
+            }
+            return NuGetManagedTemplatePackage.Deserialize(_environmentSettings, this, provider, data.MountPointUri, data.Details);
+        }
+
+        public async Task<IReadOnlyList<CheckUpdateResult>> GetLatestVersionAsync(IEnumerable<IManagedTemplatePackage> packages, IManagedTemplatePackageProvider provider, CancellationToken cancellationToken)
+        {
+            _ = packages ?? throw new ArgumentNullException(nameof(packages));
+            return await Task.WhenAll(packages.Select(async package =>
+                {
+                    if (package is NuGetManagedTemplatePackage nugetPackage)
+                    {
+                        try
+                        {
+                            (string latestVersion, bool isLatestVersion) = await _updateChecker.GetLatestVersionAsync(nugetPackage.Identifier, nugetPackage.Version, nugetPackage.NuGetSource, cancellationToken).ConfigureAwait(false);
+                            return CheckUpdateResult.CreateSuccess(package, latestVersion, isLatestVersion);
+                        }
+                        catch (PackageNotFoundException e)
+                        {
+                            return CheckUpdateResult.CreateFailure(package, InstallerErrorCode.PackageNotFound, e.Message);
+                        }
+                        catch (InvalidNuGetSourceException e)
+                        {
+                            return CheckUpdateResult.CreateFailure(package, InstallerErrorCode.InvalidSource, e.Message);
+                        }
+                        catch (Exception e)
+                        {
+                            _environmentSettings.Host.LogDiagnosticMessage($"Retrieving latest version for package {package.DisplayName} failed.", DebugLogCategory);
+                            _environmentSettings.Host.LogDiagnosticMessage($"Details:{e.ToString()}", DebugLogCategory);
+                            return CheckUpdateResult.CreateFailure(package, InstallerErrorCode.GenericError, $"Failed to check the update for the package {package.Identifier}, reason: {e.Message}");
+                        }
```

Comment:
Should we log `OperationCancelledException`s separately?

Reply:
Done.